// @nrwl/angular/src/builders/ng-packagr-lite/ng-packagr-adjustments/entry-point.di.js
"use strict";
/**
 * Wires everything together and provides it to the DI system, taking what
 * we still want from the original ng-packagr library and replacing those
 * where Nx takes over with Nx specific functions
 */
Object.defineProperty(exports, "__esModule", { value: true });
exports.NX_ENTRY_POINT_PROVIDERS = exports.NX_ENTRY_POINT_TRANSFORM = exports.NX_ENTRY_POINT_TRANSFORM_TOKEN = void 0;
const compile_ngc_di_1 = require("ng-packagr/lib/ng-package/entry-point/compile-ngc.di");
const write_bundles_1 = require("./write-bundles");
const write_package_di_1 = require("ng-packagr/lib/ng-package/entry-point/write-package.di");
const injection_js_1 = require("injection-js");
const transform_di_1 = require("ng-packagr/lib/graph/transform.di");
const entry_point_transform_1 = require("ng-packagr/lib/ng-package/entry-point/entry-point.transform");
exports.NX_ENTRY_POINT_TRANSFORM_TOKEN = new injection_js_1.InjectionToken(`nx.v1.entryPointTransform`);
exports.NX_ENTRY_POINT_TRANSFORM = transform_di_1.provideTransform({
    provide: exports.NX_ENTRY_POINT_TRANSFORM_TOKEN,
    useFactory: entry_point_transform_1.entryPointTransformFactory,
    deps: [
        compile_ngc_di_1.COMPILE_NGC_TOKEN,
        write_bundles_1.NX_WRITE_BUNDLES_TRANSFORM_TOKEN,
        write_package_di_1.WRITE_PACKAGE_TRANSFORM_TOKEN,
    ],
});
exports.NX_ENTRY_POINT_PROVIDERS = [
    exports.NX_ENTRY_POINT_TRANSFORM,
    compile_ngc_di_1.COMPILE_NGC_TRANSFORM,
    write_bundles_1.NX_WRITE_BUNDLES_TRANSFORM,
    write_package_di_1.WRITE_PACKAGE_TRANSFORM,
];
