// @nrwl/angular/src/schematics/storybook-configuration/configuration.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.storybookConfigurationGenerator = void 0;
const schematics_1 = require("@angular-devkit/schematics");
const ngcli_adapter_1 = require("@nrwl/devkit/ngcli-adapter");
function default_1(schema) {
    if (schema.generateCypressSpecs && !schema.generateStories) {
        throw new Error('Cannot set generateCypressSpecs to true when generateStories is set to false.');
    }
    return schematics_1.chain([
        schematics_1.externalSchematic('@nrwl/storybook', 'configuration', {
            name: schema.name,
            uiFramework: '@storybook/angular',
            configureCypress: schema.configureCypress,
            linter: schema.linter,
        }),
        schema.generateStories ? generateStories(schema) : schematics_1.noop(),
    ]);
}
exports.default = default_1;
function generateStories(schema) {
    return (tree, context) => {
        return schematics_1.schematic('stories', {
            name: schema.name,
            generateCypressSpecs: schema.configureCypress && schema.generateCypressSpecs,
        });
    };
}
exports.storybookConfigurationGenerator = ngcli_adapter_1.wrapAngularDevkitSchematic('@nrwl/angular', 'storybook-configuration');
